### ISSUE TITLE
vipsthumbnail problem with processing svg and heic

### ISSUE THREAD

**Issue by andy250** [libvips/libvips#2679]
I am trying to create thumbnails from multiple file formats with one command. The problem is that I don't know how exactly to "enable" specific formats with libvips or specifically vipsthumbnail. On Windows I am able to convert all standard file types except for HEIC and on Ubuntu I have trouble loading SVG and HEIC. What I did - Ubuntu 20.04:
```
$ sudo apt-get update
$ sudo apt-get install libvips libvips-dev libvips-tools
$ vipsthumbnail --vips-version
libvips 8.9.1-Sun Feb 23 08:51:26 UTC 2020
```

Then trying:
`$ vipsthumbnail file.svg -o x.jpg`
```
vipsthumbnail: unable to thumbnail gallardo.svg
VipsForeignLoad: "gallardo.svg" is not a known file format
```
The above command with the same svg file works on Windows.

I read this: https://github.com/libvips/libvips/issues/531 --> loading svg depends on librsvg2, so I tried `sudo apt-get install librsvg2-bin` (even though I thought all the libraries should be installed automatically with libvips-tools) but the problem remains. I might be missing sth fundamental regarding installing and enabling those 3rd party libraries to make vipsthumbnail work. I could not find any detailed info about this in official docs.

A few notes:
* same issue happens with HEIC files - but what is different with HEIC is that it fails both on Windows and Ubuntu
* should I maybe use some format-specific commands to load svg/heic and other less popular formats?

**Comment by jcupitt** [comment]
Hi @andy250,

SVG should work, that's curious. I tried in docker:

```
$ docker run --rm -it -v $PWD:/pics ubuntu:20.04
root@97a505a6105a:/# apt-get update
...
root@97a505a6105a:/# apt-get install libvips libvips-dev libvips-tools
... huge installation log
root@97a505a6105a:/# vips svgload
load SVG with rsvg
usage:
   svgload filename out [--option-name option-value ...]
where:
   filename     - Filename to load from, input gchararray
   out          - Output image, output VipsImage
optional arguments:
...
```

So it should be able to thumbnail SVG files. I tried with a test image:

```
root@97a505a6105a:/# cd /pics
root@97a505a6105a:/pics# vipsthumbnail lion.svg
root@97a505a6105a:/pics# vipsheader tn_lion.jpg 
tn_lion.jpg: 128x128 uchar, 3 bands, srgb, jpegload
```

I would guess your specific SVG is failing to be recognised by the libvips 8.9 SVG loader -- we've fixed a few problems with SVG detection in the last few years.

It's certainly too old for HEIF. I tried:

```
root@97a505a6105a:/pics# vips heifload
vips: unknown action "heifload"
```

When you compile libvips it searches your system for a large range of libraries (eg. libheif, etc.) and, if they are present, builds support for those file formats. If the libraries were not present when libvips was compiled, there's no way to add them.

libvips 8.12 has a revised plugin system that does let you add file formats at runtime, but your libvips is too old for that, of course.

Anyway, tldr, you need to either build your own libvips (it's not too difficult), or update your ubuntu.

**Comment by andy250** [comment]
Thanks @jcupitt for such a swift and detailed response! This generates a few more questsions:
- let's say I want to install latest and greatest libvips 8.12 - can I do it (avoid apt-get I guess)? Or is updating Ubuntu the only way to go? I planned to use Azure VMs and AFAIK I can't upgrade the OS there. Also I prefer to install required tools (like libvips) with scripts on all my servers, rather than building and distributing on my own. Just figuring out my options here (I am at the RnD stage of a project).
- "too old for HEIF" --> would libvips 8.12 handle HEIF files out of the box?

**Comment by jcupitt** [comment]
Hi again, yes, I guess there's something funny about your HEIC image. Could you share it somehow? Dropbox, perhaps? My personal email is jcupitt@gmail.com if it's confidential, or just post a link here.

Regarding deployment, you can deploy via docker on azure, so you're not tied to the platform OS. It's a very convenient way to build and maintain custom binaries. I have a few samples here:

https://github.com/jcupitt/docker-builds

This is very close to what you want:

https://github.com/jcupitt/docker-builds/blob/master/libvips-pdfium-ubuntu20.04/Dockerfile

It builds against pdfium as well, which might be handy. You'd probably want to adjust the other dependencies. That actually uses 20.04 as a base, so you could run the binaries natively if you want. Though running in a container adds some security, so it might be preferable.

The great advantage of building your own libvips binary is that you can remove things you don't need. The ubuntu libvips includes almost everything (this is why it's such a crazy download size), and some of the load libraries (eg. openslide), though very useful for some cases, are not well maintained or tested. I expect they could be exploited rather easily by an attacker.

What implementation language are you using? Does that need to run on the same platform?

**Comment by andy250** [comment]
Thanks! I'll take a look at building it myself with the dockerfile you linked.

I checked the two HEIC files again - actually they fail in a different way than the svg on Ubuntu. Here is the output:

```
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.743: error in tile 0 x 0
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.743: error in tile 0 x 0
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.743: error in tile 0 x 0
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.743: error in tile 0 x 10
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.759: error in tile 0 x 20
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.759: error in tile 0 x 30
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.759: error in tile 0 x 40
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.759: error in tile 0 x 50
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.759: error in tile 0 x 60
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.775: error in tile 0 x 70
(vipsthumbnail:636): VIPS-WARNING **: 21:59:49.775: error in tile 0 x 80
vipsthumbnail: unable to thumbnail crowd_1440x960.heic
crowd_1440x960.heic: bad seek to 130366
heif: Unsupported feature: Unsupported codec (4.3000)
```
So yes, it seems to be an issue with the file format/codec.

Files: https://1drv.ms/u/s!ArWg6Y577P00hIomuJNZdhFeYeZNKA?e=OrIfCJ

Implementation: .net 6 and running vipsthumbnail directly from the app via [NetVips](https://github.com/kleisauke/net-vips). Target platform of the app is Ubuntu.

**Comment by jcupitt** [comment]
Your files are working for me, fwiw:

```
$ vipsheader sample1.heic 
sample1.heic: 1440x960 uchar, 3 bands, srgb, heifload
$ vips copy sample1.heic x.jpg
$ vipsthumbnail sample1.heic 
```

Using libheif/de265, as bundled with ubuntu 21.10.

**Comment by jcupitt** [comment]
> So yes, it seems to be an issue with the file format/codec.

Ah I always forget. HEIC decode is patented, so netvips can't include support by default in binaries it distributes. You can build your own libvips binary that does include HEIC support, there are notes at the end of the README:

https://github.com/libvips/build-win64-mxe#hevc-related-dependencies

And of course you'll need to be rather cautious yourself, especially if you enable HEIC encode.

**Comment by andy250** [comment]
Thank you. I just want to clarify:

> **netvips** can't include support by default in binaries it distributes. You can build your own **libvips** binary
> **netvips** does include AVIF support by default

Do you mean libvips instead of netvips here? Isn't netvips just a wrapper for executing libvips binaries? Just want to make sure which components actually I need to build myself.

**Comment by jcupitt** [comment]
netvips automatically downloads and uses precompiled libvips binaries, and those binaries don't have HEIC enabled. You'll need to build your own libvips, enable HEIC, and then drop those binaries into your netvips install.

On linux you can get netvips to use the system binary, so it's a little easier.

(or that's my understanding, ask on the netvips repo for the correct answer, of course)

**Comment by kleisauke** [comment]
The [`NetVips.Native.*`](https://www.nuget.org/packages?q=NetVips.Native) NuGet packages contains the native (i.e. unmanaged) libraries for libvips. These packages are not needed if you have libvips installed globally, which is also the reason why it can be installed optionally.

These pre-compiled libvips binaries include AVIF and SVG support by default. It does not support the patent-encumbered HEIC format, as mentioned by John.

**Comment by andy250** [comment]
Thanks @kleisauke. So my understanding is that I have 2 options to support all formats I want:
* compile libvips on my own (with svg, heic etc) and make it available globally - then use just NetVips package from nuget in .net6 code
* compile NetVips.Native on my own let's name it NetVips.Native.andy250 (include HEIC in addition to what is there already) - and then use NetVips package plus NetVips.Native.andy250 in my .net6 code - no need for globally available libvips binaries

**Comment by kleisauke** [comment]
I think approach 1 is more appropriate on Linux distros. If you go for approach 2, you need to make sure that it's not published on a public package source, as this exposes you or your employer to possible royalty payments to patent pools/holders for each copy that is "downloaded".
